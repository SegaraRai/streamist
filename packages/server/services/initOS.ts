import { configDevelopment, setOS } from '$shared/objectStorage';

switch (process.env.NODE_ENV) {
  case 'development':
    setOS(configDevelopment);
    break;

  default:
    throw new Error('unknown NODE_ENV');
}
