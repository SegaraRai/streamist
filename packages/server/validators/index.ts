import 'reflect-metadata';

export * from './albums';
export * from './artists';
export * from './auth';
export * from './coArtists';
export * from './images';
export * from './playlists';
export * from './sources';
export * from './tracks';
