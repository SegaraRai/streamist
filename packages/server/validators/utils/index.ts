export * from './id';
export * from './region';
export * from './optional';
