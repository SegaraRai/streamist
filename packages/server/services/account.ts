import type { User } from '@prisma/client';
import { client } from '$/db/lib/client';
import { HTTPError } from '$/utils/httpError';
import type { IAccountUpdateData } from '$/validators';
import { calcPasswordHashAsync, verifyPasswordHashAsync } from './password';

export async function userUpdate(
  userId: string,
  data: IAccountUpdateData
): Promise<User> {
  const { displayName, region, username } = data;

  const requiresCurrentPassword = username != null || data.password != null;
  if (requiresCurrentPassword) {
    if (!data.currentPassword) {
      throw new HTTPError(
        400,
        'currentPassword is required for updating username and password'
      );
    }

    const user = await client.user.findUnique({
      where: { id: userId },
      select: {
        password: true,
      },
    });
    if (!user) {
      throw new HTTPError(404, 'User not found');
    }

    const isOk = verifyPasswordHashAsync(data.currentPassword, user.password);
    if (!isOk) {
      throw new HTTPError(401, 'currentPassword does not match');
    }
  }

  const passwordHash =
    data.password != null
      ? await calcPasswordHashAsync(data.password)
      : undefined;

  const user = await client.user.update({
    where: { id: userId },
    data: {
      username,
      password: passwordHash,
      displayName,
      region,
      updatedAt: Date.now(),
    },
  });

  return user;
}
