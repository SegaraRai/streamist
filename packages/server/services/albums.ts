import { Album, Prisma } from '@prisma/client';
import { generateArtistId } from '$shared-server/generateId';
import { emptyToNull } from '$shared/transform';
import { dbAlbumMoveImageBefore, dbAlbumRemoveImageTx } from '$/db/album';
import { dbArtistGetOrCreateByNameTx } from '$/db/artist';
import { client } from '$/db/lib/client';
import { dbCoArtistMergeTx } from '$/db/lib/coArtist';
import { dbImageDeleteByImageOrderTx, dbImageDeleteTx } from '$/db/lib/image';
import { dbDeletionAddTx, dbResourceUpdateTimestamp } from '$/db/lib/resource';
import { HTTPError } from '$/utils/httpError';
import { imageDeleteFilesAndSourceFiles } from './images';

export type AlbumUpdateData = Partial<
  Pick<Album, 'title' | 'titleSort' | 'notes' | 'artistId'>
> & {
  artistName?: string;
};

export interface UpdateAlbumOptions {
  readonly forceNewArtist?: boolean;
}

export async function albumUpdate(
  userId: string,
  albumId: string,
  data: AlbumUpdateData,
  { forceNewArtist = false }: UpdateAlbumOptions
): Promise<void> {
  await client.$transaction(async (txClient) => {
    const timestamp = Date.now();

    const albumCount = await txClient.album.count({
      where: {
        id: albumId,
        userId,
      },
    });
    if (!albumCount) {
      throw new HTTPError(404, `album ${albumId} not found`);
    }

    // create artist
    let artistId: string | undefined;
    if (data.artistId) {
      const artist = await txClient.artist.findFirst({
        where: {
          id: data.artistId,
          userId,
        },
      });
      if (!artist) {
        throw new HTTPError(400, `Artist ${data.artistId} not found`);
      }
      artistId = artist.id;
    } else if (data.artistName) {
      const artist = forceNewArtist
        ? await txClient.artist.create({
            data: {
              id: await generateArtistId(),
              name: data.artistName,
              userId,
              createdAt: timestamp,
              updatedAt: timestamp,
            },
          })
        : await dbArtistGetOrCreateByNameTx(txClient, userId, data.artistName);
      artistId = artist.id;
    }

    await txClient.album.updateMany({
      where: {
        id: albumId,
        userId,
      },
      data: {
        title: data.title,
        titleSort: emptyToNull(data.titleSort),
        notes: data.notes,
        artistId,
        updatedAt: timestamp,
      },
    });
  });

  await dbResourceUpdateTimestamp(userId);
}

export async function albumMerge(
  userId: string,
  albumId: string,
  toAlbumId: string
): Promise<void> {
  await client.$transaction(async (txClient): Promise<void> => {
    const timestamp = Date.now();

    const album = await txClient.album.findFirst({
      where: {
        id: albumId,
        userId,
      },
    });

    if (!album) {
      throw new HTTPError(404, `album ${albumId} not found`);
    }

    const toAlbum = await txClient.album.findFirst({
      where: {
        id: toAlbumId,
        userId,
      },
    });

    if (!toAlbum) {
      throw new HTTPError(404, `album ${toAlbumId} not found`);
    }

    await txClient.track.updateMany({
      where: {
        albumId,
        userId,
      },
      data: {
        albumId: toAlbumId,
        updatedAt: timestamp,
      },
    });

    await dbCoArtistMergeTx<typeof Prisma.AlbumCoArtistScalarFieldEnum>(
      txClient,
      Prisma.ModelName.AlbumCoArtist,
      'albumId',
      userId,
      albumId,
      toAlbumId,
      timestamp
    );

    await txClient.aAlbumImage.updateMany({
      where: {
        x: albumId,
        userId,
      },
      data: {
        x: toAlbumId,
      },
    });

    const updated = await txClient.album.updateMany({
      where: {
        id: toAlbumId,
        imageOrder: toAlbum.imageOrder,
        userId,
      },
      data: {
        updatedAt: timestamp,
        imageOrder: toAlbum.imageOrder + album.imageOrder,
      },
    });
    if (!updated.count) {
      throw new HTTPError(409, `album ${toAlbumId} changed during merge`);
    }

    const deleted = await txClient.album.deleteMany({
      where: {
        id: albumId,
        imageOrder: album.imageOrder,
        userId,
      },
    });
    if (!deleted.count) {
      throw new HTTPError(409, `album ${albumId} changed during merge`);
    }

    await dbDeletionAddTx(txClient, userId, 'album', albumId);
  });

  await dbResourceUpdateTimestamp(userId);
}

/**
 * 指定されたアルバムが参照されていない場合に削除する
 *
 * @description
 * - 確認する参照元: Track
 * - 同時に削除するもの: AlbumCoArtist (via CASCADE), Image, ImageFile, 画像ファイル(S3)
 * ! 本メソッドは更新日時を更新する（削除時のみ） \
 * ! 本メソッドは削除記録を追加する（削除時のみ） \
 * @param userId
 * @param albumId
 * @returns 削除されたらアルバムのアーティストID、まだ参照されている場合は`false`
 */
export async function albumDeleteIfUnreferenced(
  userId: string,
  albumId: string,
  skipUpdateTimestamp = false
): Promise<string | false> {
  // Trackから参照されていないか確認
  {
    const query = { where: { albumId, userId } } as const;
    const count = await client.track.count(query);
    if (count > 0) {
      return false;
    }
  }

  const [artistId, images] = await client.$transaction(async (txClient) => {
    // delete images
    const album = await txClient.album.findFirst({
      where: {
        id: albumId,
        userId,
      },
      select: {
        artistId: true,
        imageOrder: true,
      },
    });
    if (!album) {
      throw new HTTPError(404, `Album ${albumId} not found`);
    }

    const images = await dbImageDeleteByImageOrderTx(
      txClient,
      userId,
      album.imageOrder
    );

    // delete album
    // * Album id is referenced from: Track, AlbumCoArtist, Image (implicit m:n)
    // AlbumCoArtist will be cascade deleted (Deletion of AlbumCoArtist is not recorded, therefore the client must synchronize AlbumCoArtist based on the Deletion of Album)
    // TODO(db): set ON DELETE RESTRICT for Track and Image (implicit m:n) table
    const result = await txClient.album.deleteMany({
      where: {
        id: albumId,
        artistId: album.artistId,
        imageOrder: album.imageOrder,
        userId,
      },
    });
    if (result.count === 0) {
      throw new HTTPError(
        409,
        `Images are modified during deletion of album ${albumId}`
      );
    }

    await dbDeletionAddTx(txClient, userId, 'album', albumId);

    return [album.artistId, images] as const;
  });

  await imageDeleteFilesAndSourceFiles(userId, images, true);

  // update resource timestamp
  if (!skipUpdateTimestamp) {
    await dbResourceUpdateTimestamp(userId);
  }

  return artistId;
}

export async function albumImageMoveBefore(
  userId: string,
  albumId: string,
  imageId: string,
  referenceImageId: string | undefined
): Promise<void> {
  await dbAlbumMoveImageBefore(userId, albumId, imageId, referenceImageId);
  await dbResourceUpdateTimestamp(userId);
}

export async function albumImageDelete(
  userId: string,
  albumId: string,
  imageIds: string | readonly string[]
): Promise<void> {
  const images = await client.$transaction(async (txClient) => {
    await dbAlbumRemoveImageTx(txClient, userId, albumId, imageIds);
    return dbImageDeleteTx(txClient, userId, imageIds);
  });
  await imageDeleteFilesAndSourceFiles(userId, images, true);
  await dbResourceUpdateTimestamp(userId);
}
