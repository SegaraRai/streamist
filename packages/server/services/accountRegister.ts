import type { User } from '@prisma/client';
import { generateUserId } from '$shared-server/generateId';
import { isReservedUsername } from '$shared-server/reservedUsernames';
import { INITIAL_PLAN } from '$shared/config';
import { client } from '$/db/lib/client';
import { dbGetTimestamp } from '$/db/lib/timestamp';
import { calcPasswordHashAsync } from '$/services/password';
import { HTTPError } from '$/utils/httpError';
import type { IAccountCreateData } from '$/validators';

export async function userDoesExist(username: string): Promise<boolean> {
  if (isReservedUsername(username)) {
    return true;
  }

  const count = await client.user.count({
    where: {
      username: String(username),
    },
  });

  return count !== 0;
}

export async function userCreate(data: IAccountCreateData): Promise<User> {
  const { displayName, region, username } = data;

  if (isReservedUsername(username)) {
    throw new HTTPError(400, 'username is reserved');
  }

  const passwordHash = await calcPasswordHashAsync(data.password);

  const timestamp = dbGetTimestamp();

  const userId = await generateUserId();

  const user = await client.$transaction(async (txClient): Promise<User> => {
    const user = await txClient.user.create({
      data: {
        id: userId,
        username,
        password: passwordHash,
        displayName,
        region,
        plan: INITIAL_PLAN,
        createdAt: timestamp,
        updatedAt: timestamp,
      },
    });

    await txClient.resourceUpdate.create({
      data: {
        userId,
        updatedAt: timestamp,
      },
    });

    return user;
  });

  return user;
}
