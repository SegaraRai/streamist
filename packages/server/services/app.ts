import '$/services/initOS';
import '$/services/initCredentials';

import closeWithGrace from 'close-with-grace';
import Fastify, { FastifyServerFactory } from 'fastify';
import cors from 'fastify-cors';
import helmet from 'fastify-helmet';
import fastifyJwt from 'fastify-jwt';
import server from '$/$server';
import { devCDN } from '$/services/dev';
import { API_BASE_PATH, SECRET_API_JWT_SECRET } from '$/services/env';
import { fastPlainToInstance } from '$/services/fastClassTransformer';
import { transcoderCallback } from '$/services/transcoderCallback';

export const init = (serverFactory?: FastifyServerFactory) => {
  const app = Fastify({
    serverFactory,
    logger: {
      prettyPrint:
        process.env.NODE_ENV === 'development'
          ? {
              translateTime: 'HH:MM:ss Z',
              ignore: 'hostname',
            }
          : false,
    },
  });

  app.register(helmet, {
    contentSecurityPolicy: {
      useDefaults: false,
      directives: {
        // as this is an API server, we don't have to fetch any resources
        defaultSrc: ["'none'"],
      },
    },
  });

  app.register(transcoderCallback);

  if (process.env.NODE_ENV === 'development') {
    app.register(devCDN, {
      prefix: '/dev/cdn',
    });
  }

  // NOTE: not setting custom error handler as fastify's default one works fine
  // TODO(prod): should be set in production to collect errors and send them to sentry or something
  app.register(
    (fastify, _options, done): void => {
      if (process.env.NODE_ENV === 'development') {
        fastify.register(cors, {
          origin: true,
        });
      } else {
        // TODO(prod): set allowed origins for production and staging environments
        fastify.register(cors);
      }
      fastify.register(fastifyJwt, { secret: SECRET_API_JWT_SECRET });
      server(fastify, {
        plainToInstance: fastPlainToInstance,
      });
      done();
    },
    {
      prefix: API_BASE_PATH,
    }
  );

  const closeListeners = closeWithGrace(
    { delay: 10000 },
    ({ err }, callback): void => {
      if (err) {
        app.log.error(err);
      }
      app.log.info('shutting down worker');
      app.close().finally((): void => {
        callback();
      });
    }
  );

  app.addHook('onClose', (_instance, done): void => {
    closeListeners.uninstall();
    done();
  });

  return app;
};
