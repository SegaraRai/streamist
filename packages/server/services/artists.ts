import { Prisma } from '@prisma/client';
import { dbArtistMoveImageBefore, dbArtistRemoveImageTx } from '$/db/artist';
import { client } from '$/db/lib/client';
import { dbCoArtistMergeArtistTx } from '$/db/lib/coArtist';
import { dbImageDeleteByImageOrderTx, dbImageDeleteTx } from '$/db/lib/image';
import { dbDeletionAddTx, dbResourceUpdateTimestamp } from '$/db/lib/resource';
import { dbGetTimestamp } from '$/db/lib/timestamp';
import { imageDeleteFilesAndSourceFiles } from '$/services/images';
import { HTTPError } from '$/utils/httpError';
import type { IArtistUpdateData } from '$/validators';

export async function artistUpdate(
  userId: string,
  artistId: string,
  data: IArtistUpdateData
): Promise<void> {
  const { description, name, nameSort } = data;

  if (
    description === undefined &&
    name === undefined &&
    nameSort === undefined
  ) {
    return;
  }

  const artist = await client.artist.updateMany({
    where: {
      id: artistId,
      userId,
    },
    data: {
      name,
      nameSort,
      description,
      updatedAt: dbGetTimestamp(),
    },
  });
  if (artist.count === 0) {
    throw new HTTPError(404, `artist ${artistId} not found`);
  }

  await dbResourceUpdateTimestamp(userId);
}

export async function artistMerge(
  userId: string,
  artistId: string,
  toArtistId: string
): Promise<void> {
  await client.$transaction(async (txClient): Promise<void> => {
    const timestamp = dbGetTimestamp();

    const artist = await txClient.artist.findFirst({
      where: {
        id: artistId,
        userId,
      },
      select: {
        imageOrder: true,
      },
    });

    if (!artist) {
      throw new HTTPError(404, `artist ${artistId} not found`);
    }

    const toArtist = await txClient.artist.findFirst({
      where: {
        id: toArtistId,
        userId,
      },
      select: {
        imageOrder: true,
      },
    });

    if (!toArtist) {
      throw new HTTPError(404, `artist ${toArtistId} not found`);
    }

    const query = {
      where: {
        artistId,
        userId,
      },
      data: {
        artistId: toArtistId,
        updatedAt: timestamp,
      },
    } as const;

    await txClient.track.updateMany(query);
    await txClient.album.updateMany(query);
    await dbCoArtistMergeArtistTx<typeof Prisma.TrackCoArtistScalarFieldEnum>(
      txClient,
      Prisma.ModelName.TrackCoArtist,
      'trackId',
      userId,
      artistId,
      toArtistId,
      timestamp
    );
    await dbCoArtistMergeArtistTx<typeof Prisma.AlbumCoArtistScalarFieldEnum>(
      txClient,
      Prisma.ModelName.AlbumCoArtist,
      'albumId',
      userId,
      artistId,
      toArtistId,
      timestamp
    );

    await txClient.artist.updateMany({
      where: {
        id: toArtistId,
        userId,
      },
      data: {
        updatedAt: timestamp,
      },
    });

    const deleted = await txClient.artist.deleteMany({
      where: {
        id: artistId,
        userId,
      },
    });
    if (!deleted.count) {
      throw new HTTPError(409, `artist ${artistId} deleted during merge`);
    }

    await dbDeletionAddTx(txClient, userId, 'artist', artistId);
  });

  await dbResourceUpdateTimestamp(userId);
}

/**
 * 指定されたアーティストが参照されていない場合に削除する
 *
 * @description
 * - 確認する参照元: Album, Track, AlbumCoArtist, TrackCoArtist
 * - 同時に削除するもの: Image, ImageFile, 画像ファイル(S3)
 * ! 本メソッドは更新日時を更新する（削除時のみ） \
 * ! 本メソッドは削除記録を追加する（削除時のみ）
 * @param userId
 * @param artistId
 * @returns 削除されたら`true`、まだ参照されている場合は`false`
 */
export async function artistDeleteIfUnreferenced(
  userId: string,
  artistId: string,
  skipUpdateTimestamp = false
): Promise<boolean> {
  // Album, Track, AlbumCoArtist, TrackCoArtistから参照されていないか確認
  // 短絡評価を活用するため、より頻繁に参照されていそうなもの順にしておく
  {
    const query = { where: { artistId, userId } } as const;
    const count =
      (await client.album.count(query)) ||
      (await client.track.count(query)) ||
      (await client.albumCoArtist.count(query)) ||
      (await client.trackCoArtist.count(query));
    if (count > 0) {
      return false;
    }
  }

  const images = await client.$transaction(async (txClient) => {
    // delete images
    const artist = await txClient.artist.findFirst({
      where: {
        id: artistId,
        userId,
      },
      select: {
        imageOrder: true,
      },
    });
    if (!artist) {
      throw new HTTPError(404, `Artist ${artistId} not found`);
    }

    const images = await dbImageDeleteByImageOrderTx(
      txClient,
      userId,
      artist.imageOrder
    );

    // delete artist
    // * Artist id is referenced from: Album, AlbumCoArtist, Track, TrackCoArtist, Image (implicit m:n)
    const result = await txClient.artist.deleteMany({
      where: {
        id: artistId,
        imageOrder: artist.imageOrder,
        userId,
      },
    });
    if (result.count === 0) {
      throw new HTTPError(
        409,
        `Images are modified during deletion of artist ${artistId}`
      );
    }
    await dbDeletionAddTx(txClient, userId, 'artist', artistId);

    return images;
  });

  await imageDeleteFilesAndSourceFiles(userId, images, true);

  // update resource timestamp
  if (!skipUpdateTimestamp) {
    await dbResourceUpdateTimestamp(userId);
  }

  return true;
}

export async function artistImageMoveBefore(
  userId: string,
  artistId: string,
  imageId: string,
  referenceImageId: string | undefined
): Promise<void> {
  await dbArtistMoveImageBefore(userId, artistId, imageId, referenceImageId);
  await dbResourceUpdateTimestamp(userId);
}

export async function artistImageDelete(
  userId: string,
  artistId: string,
  imageIds: string | readonly string[]
): Promise<void> {
  const images = await client.$transaction(async (txClient) => {
    await dbArtistRemoveImageTx(txClient, userId, artistId, imageIds);
    return dbImageDeleteTx(txClient, userId, imageIds);
  });
  await imageDeleteFilesAndSourceFiles(userId, images, true);
  await dbResourceUpdateTimestamp(userId);
}
