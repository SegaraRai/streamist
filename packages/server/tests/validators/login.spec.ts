import '$/services/initOS';

import { describe, expect, it } from 'vitest';
import { VAuthBodyPassword } from '$/validators';
import { validate } from './utils';

describe('VAuthBodyWrapper', () => {
  it('should accept correct payloads', async () => {
    await expect(
      validate(VAuthBodyPassword, {
        grant_type: 'password',
        username: 'test_id',
        password: 'test_password',
      })
    ).resolves.toBeUndefined();
  });

  it('should reject malformed payloads', async () => {
    // invalid grant_type
    await expect(
      validate(VAuthBodyPassword, {
        grant_type: 'foo' as any,
        username: 'test_id',
        password: 'test_password',
      })
    ).rejects.toMatchInlineSnapshot(`
      [
        ValidationError {
          "children": [],
          "constraints": {
            "equals": "grant_type must be equal to password",
          },
          "property": "grant_type",
          "value": "foo",
        },
      ]
    `);

    // missing grant_type
    await expect(
      validate(VAuthBodyPassword, {
        username: 'test_id',
        password: 'test_password',
      } as any)
    ).rejects.toMatchInlineSnapshot(`
      [
        ValidationError {
          "children": [],
          "constraints": {
            "equals": "grant_type must be equal to password",
          },
          "property": "grant_type",
          "value": undefined,
        },
      ]
    `);

    // invalid username
    await expect(
      validate(VAuthBodyPassword, {
        grant_type: 'password',
        username: { toString: 'test_id' },
        password: 'test_password',
      } as any)
    ).rejects.toMatchSnapshot();

    // invalid password
    await expect(
      validate(VAuthBodyPassword, {
        grant_type: 'password',
        username: 'test_id',
        password: 123 as any,
      } as any)
    ).rejects.toMatchSnapshot();

    // missing password
    await expect(
      validate(VAuthBodyPassword, {
        grant_type: 'password',
        username: 'test_id',
      } as any)
    ).rejects.toMatchSnapshot();
  });
});
