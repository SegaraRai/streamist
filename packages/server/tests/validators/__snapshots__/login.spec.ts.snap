// Vitest Snapshot v1

exports[`VAuthBodyWrapper > should reject malformed payloads 1`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "equals": "grant_type must be equal to password",
    },
    "property": "grant_type",
    "value": "foo",
  },
]
`;

exports[`VAuthBodyWrapper > should reject malformed payloads 2`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "equals": "grant_type must be equal to password",
    },
    "property": "grant_type",
    "value": undefined,
  },
]
`;

exports[`VAuthBodyWrapper > should reject malformed payloads 3`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "username must be a string",
      "matches": "username must match /^[a-z0-9_]+$/ regular expression",
      "maxLength": "username must be shorter than or equal to 32 characters",
      "minLength": "username must be longer than or equal to 3 characters",
    },
    "property": "username",
    "value": {},
  },
]
`;

exports[`VAuthBodyWrapper > should reject malformed payloads 4`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isString": "password must be a string",
      "matches": "password must match /^[\\\\w!\\"#$%&'()*+,./:;<=>?@[\\\\\\\\\\\\]^\`{|}~-]+$/ regular expression",
      "maxLength": "password must be shorter than or equal to 1024 characters",
      "minLength": "password must be longer than or equal to 8 characters",
    },
    "property": "password",
    "value": 123,
  },
]
`;

exports[`VAuthBodyWrapper > should reject malformed payloads 5`] = `
[
  ValidationError {
    "children": [],
    "constraints": {
      "isNotEmpty": "password should not be empty",
      "isString": "password must be a string",
      "matches": "password must match /^[\\\\w!\\"#$%&'()*+,./:;<=>?@[\\\\\\\\\\\\]^\`{|}~-]+$/ regular expression",
      "maxLength": "password must be shorter than or equal to 1024 characters",
      "minLength": "password must be longer than or equal to 8 characters",
    },
    "property": "password",
    "value": undefined,
  },
]
`;
